lib/iteration.rb

```diff
@@ -0,0 +1,109 @@
+def benchmark(num = 1)
+  start = Time.now
+  num.times { yield }
+  elapsed_time = (Time.now - start) / num
+end
+
+def race(procs, num = 1)
+  arr = []
+  procs.each do |proc|
+    runtime = benchmark(num)
+    arr << runtime
+  end
+  index = arr.index(arr.max)
+  procs[index]
+end
+
+def eval_block(*args, &prc)
+  if prc.nil?
+    "NO BLOCK GIVEN"
+  else
+    p yield *args
+  end
+end
+
+def factors(num)
+  (1..num).select { |n| num % n == 0 }
+end
+
+class Array
+  def bubble_sort!(&prc)
+    sorted = false
+    until sorted
+      sorted = true
+      (0..self.length - 2).each do |idx|
```

Comment:
Good

On Tue, May 31, 2016 at 12:33 PM, Jeff Liu <notifications@github.com> wrote:

> In lib/iteration.rb
> <https://github.com/emperorliu/test-first-ruby/pull/4#discussion_r65247828>
> :
>
> > +    "NO BLOCK GIVEN"
> > +  else
> > +    p yield *args
> > +  end
> > +end
> > +
> > +def factors(num)
> > +  (1..num).select { |n| num % n == 0 }
> > +end
> > +
> > +class Array
> > +  def bubble_sort!(&prc)
> > +    sorted = false
> > +    until sorted
> > +      sorted = true
> > +      (0..self.length - 2).each do |idx|
>
> Because in an array like [ 1, 3, 4, 5, 6 ] which has 5 elements, I only
> want to sort from position 0-3, which is basically the numbers 1, 3, 4, 5.
> Once I compare 5 with 6, I don't need to move forward anymore because I
> don't have any thing to compare the last element with.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/emperorliu/test-first-ruby/pull/4/files/dd2b0b193a8a825ec50c28358bc98518685c32d5#r65247828>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABtEnKyoZat_lmfx_NLv2YakiD_YuSz5ks5qHIzsgaJpZM4IqHIJ>
> .
>